matchbox/provider_test.go

```diff
@@ -0,0 +1,251 @@
+package matchbox
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"net"
+	"testing"
+
+	"github.com/coreos/matchbox/matchbox/rpc"
+	"github.com/coreos/matchbox/matchbox/server"
+	"github.com/coreos/matchbox/matchbox/storage"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+var providers = map[string]terraform.ResourceProvider{
+	"matchbox": Provider(),
+}
+
+func TestProvider(t *testing.T) {
+	if err := Provider().(*schema.Provider).InternalValidate(); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+}
+
+type MatchboxFixtureServer struct {
+	Store    storage.Store
+	Listener net.Listener
+}
+
+func NewMatchboxFixtureServer(s storage.Store) *MatchboxFixtureServer {
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		panic(fmt.Errorf("failed to start listener: %v", err))
+	}
+
+	return &MatchboxFixtureServer{
+		Store:    s,
+		Listener: l,
+	}
+}
+
+func (s *MatchboxFixtureServer) Start() {
+	cfg, err := testServerTLSConfig()
+	if err != nil {
+		panic(fmt.Errorf("Invalid TLS credentials: %v", err))
+	}
+
+	srv := server.NewServer(&server.Config{s.Store})
```

Comment:
Again, let's keep it similar to the real server. Listener need not be in the struct and there is a utility for constructing the server's TLS Config.

```
func (s *MatchboxFixtureServer) Start() {
    svr := server.NewServer(&server.Config{Store: s.Store})
    lis, err := net.Listen("tcp", 127.0.0.1:8081)
    if err != nil {
       panic(fmt.Printf("failed to start listening: %v", err)
    }
    tlscfg, err := s.TLSInfo.ServerConfig()
    if err != nil {
       log.Fatalf("Invalid TLS credentials: %v", err)
    }
    grpcServer := rpc.NewServer(server, tlscfg)
    grpcServer.Serve(lis)
}
```

I think the caller should run Start with a go-routine when appropriate, not within the function.